bucket-policy.go

```diff
@@ -383,90 +324,15 @@ func removeBucketPolicyStatement(statements []Statement, bucketName string, obje
 
 // Checks if an access policiy is defined for the given object prefix
 func isPolicyDefinedForObjectPrefix(statements []Statement, bucketName string, objectPrefix string) bool {
+	objectResource := awsResourcePrefix + bucketName + "/" + objectPrefix + "*"
 	for _, statement := range statements {
-		for _, resource := range statement.Resources {
-			if resource == awsResourcePrefix+bucketName+"/"+objectPrefix+"*" {
-				return true
-			}
+		if statement.Resources.Contains(objectResource) {
+			return true
 		}
 	}
 	return false
 }
 
-// Unmarshals bucket policy byte array into a structured bucket access policy.
-func unMarshalBucketPolicy(bucketPolicyBuf []byte) (BucketAccessPolicy, error) {
-	// Untyped lazy JSON struct.
-	type bucketAccessPolicyUntyped struct {
-		Version   string
-		Statement []struct {
-			Sid       string
-			Effect    string
-			Principal struct {
-				AWS json.RawMessage
-			}
-			Action    json.RawMessage
-			Resource  json.RawMessage
-			Condition map[string]map[string]string
-		}
-	}
-	var policyUntyped = bucketAccessPolicyUntyped{}
-	// Unmarshal incoming policy into an untyped structure, to be
-	// evaluated lazily later.
-	err := json.Unmarshal(bucketPolicyBuf, &policyUntyped)
-	if err != nil {
-		return BucketAccessPolicy{}, err
-	}
-	var policy = BucketAccessPolicy{}
-	policy.Version = policyUntyped.Version
-	for _, stmtUntyped := range policyUntyped.Statement {
-		statement := Statement{}
-		// These are properly typed messages.
-		statement.Sid = stmtUntyped.Sid
-		statement.Effect = stmtUntyped.Effect
-		statement.Conditions = stmtUntyped.Condition
-
-		// AWS user can have two different types, either as []string
-		// and either as regular 'string'. We fall back to doing this
-		// since there is no other easier way to fix this.
-		err = json.Unmarshal(stmtUntyped.Principal.AWS, &statement.Principal.AWS)
-		if err != nil {
-			var awsUser string
-			err = json.Unmarshal(stmtUntyped.Principal.AWS, &awsUser)
```

Comment:
`Principal.AWS` is `set.StringSet` type which is taking care of a string or string slice automatically.